zetaclient/evm_client.go

```diff
@@ -1129,3 +1247,15 @@ func (ob *EVMChainClient) GetTxID(nonce uint64) string {
 	tssAddr := ob.Tss.EVMAddress().String()
 	return fmt.Sprintf("%d-%s-%d", ob.chain.ChainId, tssAddr, nonce)
 }
+
+func (ob *EVMChainClient) GetBlockByNumberCached(blockNumber int64) (*ethtypes.Block, error) {
+	if block, ok := ob.BlockCache.Get(blockNumber); ok {
+		return block.(*ethtypes.Block), nil
+	}
+	block, err := ob.EvmClient.BlockByNumber(context.Background(), big.NewInt(blockNumber))
+	if err != nil {
+		return nil, err
+	}
+	ob.BlockCache.Add(blockNumber, block)
```

Comment:
What abount sseprating the Add and Get for blockcache  . We might need to handle it the failure for PostBlokcHeader

Reply:
let's do it in a separate PR

Reply:
I guess the `GetBlockByNumberCached` is doing its block-scanning job in a pretty relaxed way (without caring about errors). It's an okay way for the moment.